File: python/test_heritage.py
"""Tests TP1 — Héritage Personne / Etudiant / Cours.

Cycle TDD : ces tests sont écrits AVANT l'implémentation (phase RED),
puis les classes sont implémentées jusqu'à GREEN, puis refactorisées.
"""

import pytest

from cours import Cours
from etudiant import Etudiant
from personne import Personne


# ---------- Cours ----------
class TestCours:
    def test_creation_cours_valide(self):
        c = Cours("Mathématiques", "M. Dupont")
        assert c.nom_cours == "Mathématiques"
        assert c.professeur_responsable == "M. Dupont"

    def test_str_cours_contient_nom_et_prof(self):
        c = Cours("Physique", "Mme Curie")
        s = str(c)
        assert "Physique" in s
        assert "Curie" in s

    def test_cours_nom_vide_refuse(self):
        with pytest.raises(ValueError):
            Cours("", "M. Dupont")

    def test_cours_prof_vide_refuse(self):
        with pytest.raises(ValueError):
            Cours("Maths", "   ")


# ---------- Personne ----------
class TestPersonne:
    def test_creation_personne_valide(self):
        p = Personne("Alice", 25)
        assert p.nom == "Alice"
        assert p.age == 25

    def test_str_personne(self):
        p = Personne("Bob", 30)
        s = str(p)
        assert "Bob" in s
        assert "30" in s


# ---------- Etudiant : héritage ----------
class TestEtudiantHeritage:
    def test_etudiant_est_une_personne(self):
        e = Etudiant("Charlie", 20, "E001", 14.5)
        assert isinstance(e, Personne)
        assert isinstance(e, Etudiant)

    def test_etudiant_herite_attributs_personne(self):
        e = Etudiant("Charlie", 20, "E001", 14.5)
        assert e.nom == "Charlie"
        assert e.age == 20

    def test_etudiant_attributs_specifiques(self):
        e = Etudiant("Charlie", 20, "E001", 14.5)
        assert e.numero_etudiant == "E001"
        assert e.moyenne == 14.5
        assert e.liste_cours == []

    def test_ajouter_cours(self):
        e = Etudiant("Charlie", 20, "E001")
        c1 = Cours("Maths", "M. Dupont")
        c2 = Cours("Physique", "Mme Curie")
        e.ajouter_cours(c1)
        e.ajouter_cours(c2)
        assert len(e.liste_cours) == 2
        assert e.liste_cours[0].nom_cours == "Maths"

    def test_ajouter_cours_type_invalide_refuse(self):
        e = Etudiant("Charlie", 20, "E001")
        with pytest.raises(TypeError):
            e.ajouter_cours("Pas un cours")

    def test_str_etudiant_complet(self):
        e = Etudiant("Charlie", 20, "E001", 14.5)
        e.ajouter_cours(Cours("Maths", "M. Dupont"))
        s = str(e)
        assert "Charlie" in s
        assert "E001" in s
        assert "14" in s
        assert "Maths" in s

File: java/src/test/java/com/ensta/gestion/TestHeritage.java
package com.ensta.gestion;

import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests TP1 — Héritage Personne / Etudiant / Cours.
 * Cycle TDD : ces tests sont écrits AVANT l'implémentation (RED → GREEN → REFACTOR).
 */
class TestHeritage {

    @Nested
    class CoursTests {
        @Test
        void creationCoursValide() {
            Cours c = new Cours("Mathématiques", "M. Dupont");
            assertEquals("Mathématiques", c.getNomCours());
            assertEquals("M. Dupont", c.getProfesseurResponsable());
        }

        @Test
        void toStringContientNomEtProf() {
            Cours c = new Cours("Physique", "Mme Curie");
            String s = c.toString();
            assertTrue(s.contains("Physique"));
            assertTrue(s.contains("Curie"));
        }

        @Test
        void coursNomVideRefuse() {
            assertThrows(IllegalArgumentException.class,
                () -> new Cours("", "M. Dupont"));
        }

        @Test
        void coursProfVideRefuse() {
            assertThrows(IllegalArgumentException.class,
                () -> new Cours("Maths", "   "));
        }
    }

    @Nested
    class PersonneTests {
        @Test
        void creationPersonneValide() {
            Personne p = new Personne("Alice", 25);
            assertEquals("Alice", p.getNom());
            assertEquals(25, p.getAge());
        }

        @Test
        void toStringPersonne() {
            Personne p = new Personne("Bob", 30);
            String s = p.toString();
            assertTrue(s.contains("Bob"));
            assertTrue(s.contains("30"));
        }
    }

    @Nested
    class EtudiantHeritageTests {
        @Test
        void etudiantEstUnePersonne() {
            Etudiant e = new Etudiant("Charlie", 20, "E001", 14.5);
            assertInstanceOf(Personne.class, e);
            assertInstanceOf(Etudiant.class, e);
        }

        @Test
        void etudiantHeriteAttributsPersonne() {
            Etudiant e = new Etudiant("Charlie", 20, "E001", 14.5);
            assertEquals("Charlie", e.getNom());
            assertEquals(20, e.getAge());
        }

        @Test
        void etudiantAttributsSpecifiques() {
            Etudiant e = new Etudiant("Charlie", 20, "E001", 14.5);
            assertEquals("E001", e.getNumeroEtudiant());
            assertEquals(14.5, e.getMoyenne(), 0.001);
            assertTrue(e.getListeCours().isEmpty());
        }

        @Test
        void ajouterCours() {
            Etudiant e = new Etudiant("Charlie", 20, "E001");
            e.ajouterCours(new Cours("Maths", "M. Dupont"));
            e.ajouterCours(new Cours("Physique", "Mme Curie"));
            assertEquals(2, e.getListeCours().size());
            assertEquals("Maths", e.getListeCours().get(0).getNomCours());
        }

        @Test
        void ajouterCoursNullRefuse() {
            Etudiant e = new Etudiant("Charlie", 20, "E001");
            assertThrows(IllegalArgumentException.class, () -> e.ajouterCours(null));
        }

        @Test
        void afficherDetailsEtudiantComplet() {
            Etudiant e = new Etudiant("Charlie", 20, "E001", 14.5);
            e.ajouterCours(new Cours("Maths", "M. Dupont"));
            String s = e.afficherDetails();
            assertTrue(s.contains("Charlie"));
            assertTrue(s.contains("E001"));
            assertTrue(s.contains("14"));
            assertTrue(s.contains("Maths"));
        }
    }
}
